### ISSUE TITLE
Consumers consume old messages

### ISSUE THREAD

**Issue by andreieuganox** [memphisdev/memphis.js#202]
**Describe the bug**
Yesterday, I had two Consumers running on the local machine. Today, I opened the laptop and restarted that consumer. Straight away, Consumers picked up **ALL** the messages processed yesterday!!!

**Extra information**
The consumer names did not change. 
Memphis Docker image was not restarted. 
Station storage used is Disk. 
No HA. 
1 Partition.
Version: 1.2.0 (Stable)

**To Reproduce**
1. Start one or several consumers;
2. Send a few messages;
3. Ack the messages by those consumers;
4. Kill consumers;
5. Wait for XXX
6. Start consumers again.
The messages will be consumed again.

**Expected behavior**
The message ack by the Consumer should **NEVER** be consumed by the same consumer again.

**Logs**
```
**00:18** 

> rnd-memphis@1.0.0 start
> node --watch --inspect=9242 ./src/index.js

(node:83113) ExperimentalWarning: Watch mode is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Debugger listening on ws://127.0.0.1:9242/2e5ec6fc-1df3-47a4-a5d0-92bbc2dd0784
For help, see: https://nodejs.org/en/docs/inspector
demo
demo2
Worker demo initiating...
Worker demo2 initiating...
Worker demo initiated with stationName: test-station and consumerPartitionKey: ui
Worker demo2 initiated with stationName: test-station and consumerPartitionKey: ui2
{ occupation: 'Software Engineer', interests: 'Sports,' }
{ occupation: 'Software Engineer', interests: 'Sports,' }
{ name: 'John', interests: 'Sports,', occupation: 'Software Engineer' }
{ name: 'John', interests: 'Sports,', occupation: 'Software Engineer' }
```

And
```
**10:35**

> rnd-memphis@1.0.0 start
> node --watch --inspect=9242 ./src/index.js

(node:90570) ExperimentalWarning: Watch mode is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Debugger listening on ws://127.0.0.1:9242/51488391-a036-4803-8fbd-dbf1234ecdae
For help, see: https://nodejs.org/en/docs/inspector
demo
demo2
Worker demo initiating...
Worker demo2 initiating...
Worker demo initiated with stationName: test-station and consumerPartitionKey: ui
{ occupation: 'Software Engineer', address: '123 Main St' }
{ city: 'New York', interests: 'Sports,', email: 'john@example.com' }
{ interests: 'Sports,', occupation: 'Software Engineer' }
{
  phone: '555-124',
  age: '25',
  city: 'New York',
  email: 'john@example.com'
}
{ address: '123 Main St', city: 'New York', name: 'John' }
{ address: '123 Main St', email: 'john@example.com', name: 'John' }
{ name: 'John', occupation: 'Software Engineer', age: '25' }
{
  interests: 'Sports,',
  age: '25',
  country: 'USA',
  email: 'john@example.com'
}
{
  phone: '555-124',
  age: '25',
  interests: 'Sports,',
  address: '123 Main St'
}
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
Worker demo2 initiated with stationName: test-station and consumerPartitionKey: ui2
{ occupation: 'Software Engineer', address: '123 Main St' }
{ city: 'New York', interests: 'Sports,', email: 'john@example.com' }
{ interests: 'Sports,', occupation: 'Software Engineer' }
{
  phone: '555-124',
  age: '25',
  city: 'New York',
  email: 'john@example.com'
}
{ address: '123 Main St', city: 'New York', name: 'John' }
{ address: '123 Main St', email: 'john@example.com', name: 'John' }
{ name: 'John', occupation: 'Software Engineer', age: '25' }
{
  interests: 'Sports,',
  age: '25',
  country: 'USA',
  email: 'john@example.com'
}
{
  phone: '555-124',
  age: '25',
  interests: 'Sports,',
  address: '123 Main St'
}
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var1' }
{ key: 'value', var: 'var3' }
{ city: 'New York', hobby: 'Cooking', interests: 'Sports,' }
{ occupation: 'Software Engineer', interests: 'Sports,' }
{ name: 'John', interests: 'Sports,', occupation: 'Software Engineer' }
{ key: 'value', var: 'var3' }
{ city: 'New York', hobby: 'Cooking', interests: 'Sports,' }
{ occupation: 'Software Engineer', interests: 'Sports,' }
{ name: 'John', interests: 'Sports,', occupation: 'Software Engineer' }
```

**Comment by yanivbh1** [comment]
Hey @andreieuganox ,
Thank you for bringing your issue up.
Adding @shay23b, @shohamroditimemphis for the help and debug.
Thanks.

**Comment by shay23b** [comment]
Hi @andreieuganox 
This is because the consumers were not active so they get deleted by design after a period of time, so when you ran it again it's like a new consumer - default is 3 hours, you can change it in the configurations tab to up to 48 hours

**Comment by yanivbh1** [comment]
@shay23b @shohamroditimemphis

**Comment by yanivbh1** [comment]
Hey @andreieuganox ,
Yes, it is. The purpose of identifiers is to persist in state.
Behind the stage, there is a metadata store that can be filled up and therefore we keep a "garbage collector" based on time as @shay23b mentioned. There is no such a thing as unlimited storage, and therefore we keep your storage capacity safe by removing some metadata after specific, defined, time.

**Comment by itajenglish** [comment]
@yanivbh1 Adding my thoughts on this as well. I think the garbage collection of consumers should be a developer opt-in as there could be many reasons why a consumer has not connected for x amount of time. In most cases it's an unwanted effect for consumers to receive messages it's already processed. The alternative is developers would have to keep track of messages already processed in the consumer in its local DB or elsewhere which could add additional load on the DB based on how many messages the consumers have to attempt to reprocess when it's reconnected.

**Comment by yanivbh1** [comment]
Hey @itajenglish,
Is 48 hours a reasonable amount of time? an alternative solution would be to set this amount of time as the default.

**Comment by itajenglish** [comment]
@yanivbh1 I think no time as a default would be better with the option to set a time. There can be any number of reasons a consumer hasn't been connected for a period of time but will be reconnected in the future. In local dev circumstances a developer could have gone on vacation and when they return to work all of their consumers would attempt to reprocess messages. 2. In production circumstances there could be a major outage that takes longer than 48 hours to resolve or if a team is rebuilding a service that won't be running while the rebuild is happening this can take much longer than 48 hours to complete. When they go to connect their new consumer it will attempt to process messages it has already consumed which would be unwanted. I think it should be left up to the developer to decide if they want consumers to be garbage collected if not connected after x time. At the very least consumers shouldn't be reprocessing messages it has already consumed unless the developer initiates that.

**Comment by itajenglish** [comment]
Any updates on this?

**Comment by yanivbh1** [comment]
@itajenglish will circle back once the roadmap and timelines for 2024 will be finalized.

**Comment by idanasulinmemphis** [comment]
Hi @andreieuganox, @itajenglish actually that a fix for that is going to be released on the coming version v1.4.2.
We fixed that so that the garbage collection will always remove disconnected consumers except 1 in each CG, so that the state will be kept between garbage collection cycles. 
So I'm closing this issue as completed
